// @nimbus/pets-front/src/components/PetForm.tsx
import { useState } from 'react';
import { Button, TextField, theme } from '@nimbus/commons';
import type { Pet, PetStatus } from '@nimbus/commons';

export type PetDraft = Omit<Pet, 'id'>;

interface Props {
  initial?: Pet;
  onSubmit: (draft: PetDraft) => void;
  onCancel: () => void;
}

const empty: PetDraft = { name: '', species: '', breed: '', age: 0, status: 'available' };

const STATUS_OPTIONS: { value: PetStatus; label: string }[] = [
  { value: 'available', label: 'En adopcion' },
  { value: 'treatment', label: 'En tratamiento' },
  { value: 'adopted', label: 'Adoptada' },
];

export function PetForm({ initial, onSubmit, onCancel }: Props) {
  const [draft, setDraft] = useState<PetDraft>(
    initial
      ? { name: initial.name, species: initial.species, breed: initial.breed, age: initial.age, status: initial.status }
      : empty
  );

  const set = (patch: Partial<PetDraft>) => setDraft((current) => ({ ...current, ...patch }));
  const valid =
    draft.name.trim().length > 0 &&
    draft.species.trim().length > 0 &&
    Number.isFinite(draft.age) &&
    draft.age >= 0;

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();

        if (valid) {
          onSubmit(draft);
        }
      }}
      style={{ display: 'grid', gap: 14 }}
    >
      <TextField label="Nombre" value={draft.name} onChange={(event) => set({ name: event.target.value })} autoFocus />
      <TextField label="Especie" value={draft.species} onChange={(event) => set({ species: event.target.value })} />
      <TextField label="Raza" value={draft.breed} onChange={(event) => set({ breed: event.target.value })} />
      <div style={{ display: 'grid', gridTemplateColumns: '1fr 1fr', gap: 12 }}>
        <TextField
          label="Edad (anios)"
          type="number"
          min={0}
          value={draft.age}
          onChange={(event) => set({ age: Number(event.target.value) })}
        />
        <label style={{ display: 'flex', flexDirection: 'column', gap: 6 }}>
          <span style={{ fontSize: 12, fontWeight: 600, color: theme.colors.textMuted, fontFamily: theme.font.family }}>Estado</span>
          <select
            value={draft.status}
            onChange={(event) => set({ status: event.target.value as PetStatus })}
            style={{
              height: 40,
              padding: '0 12px',
              borderRadius: theme.radius.sm,
              border: `1px solid ${theme.colors.border}`,
              fontSize: 14,
              fontFamily: theme.font.family,
              color: theme.colors.text,
              outline: 'none',
              background: '#fff',
            }}
          >
            {STATUS_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </select>
        </label>
      </div>
      <div style={{ display: 'flex', justifyContent: 'flex-end', gap: 10, marginTop: 4 }}>
        <Button type="button" variant="ghost" onClick={onCancel}>
          Cancelar
        </Button>
        <Button type="submit" disabled={!valid}>
          {initial ? 'Guardar cambios' : 'Registrar mascota'}
        </Button>
      </div>
      {!valid && (
        <span style={{ fontSize: 12, color: theme.colors.textMuted }}>
          Nombre, especie y una edad valida mayor o igual a 0 son obligatorios.
        </span>
      )}
    </form>
  );
}
